frontend/src/poc-code/testK8s/ws/WSFactory.ts

```diff
@@ -0,0 +1,338 @@
+      this.flushCanceler = setInterval(
+        this.flushMessageBuffer.bind(this),
+        this.options.bufferFlushInterval || 500,
+      );
+    }
+  }
+
+  private reconnect() {
+    if (this.connectionAttempt || this.state === 'destroyed') {
+      return;
+    }
+
+    let delay = 1000;
+
+    const attempt = () => {
+      if (!this.options.reconnect || this.state === 'open') {
+        clearTimeout(this.connectionAttempt);
+        this.connectionAttempt = null;
+        return;
+      }
+      if (this.options.timeout && delay > this.options.timeout) {
+        clearTimeout(this.connectionAttempt);
+        this.connectionAttempt = null;
+        this.destroy();
+        return;
+      }
+
+      this.connect();
+      delay = Math.round(Math.min(1.5 * delay, 60000));
+      this.connectionAttempt = setTimeout(attempt, delay);
+      // eslint-disable-next-line no-console
+      console.log(`attempting reconnect in ${delay / 1000} seconds...`);
+    };
+
+    this.connectionAttempt = setTimeout(attempt, delay);
+  }
+
+  private connect() {
+    this.state = 'init';
+    this.messageBuffer = [];
+
+    const url = createURL(applyConfigHost(this.options.host), this.options.path);
+    const subProtocols = applyConfigSubProtocols(
+      this.options.host ? this.options.subProtocols : undefined,
+    );
+    try {
+      this.ws = new WebSocket(url, subProtocols);
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.error('Error creating websocket:', e);
+      this.reconnect();
+      return;
+    }
+
+    this.ws.onopen = () => {
+      // eslint-disable-next-line no-console
+      console.log(`websocket open: ${this.id}`);
+      this.state = 'open';
+      this.triggerEvent('open', undefined);
+      if (this.connectionAttempt) {
+        clearTimeout(this.connectionAttempt);
+        this.connectionAttempt = null;
+      }
+    };
+    this.ws.onclose = (evt) => {
+      // eslint-disable-next-line no-console
+      console.log(`websocket closed: ${this.id}`, evt);
+      this.state = 'closed';
+      this.triggerEvent('close', evt);
+      this.reconnect();
+    };
+    this.ws.onerror = (evt) => {
+      // eslint-disable-next-line no-console
+      console.log(`websocket error: ${this.id}`);
+      this.state = 'error';
+      this.triggerEvent('error', evt);
+    };
+    this.ws.onmessage = (evt) => {
+      const msg = this.options?.jsonParse ? JSON.parse(evt.data) : evt.data;
+      // In some browsers, onmessage can fire after onclose/error. Don't update state to be incorrect.
+      if (this.state !== 'destroyed' && this.state !== 'closed') {
+        this.state = 'open';
+      }
+      this.triggerEvent('message', msg);
+    };
+  }
+
+  /**
+   * Invoke all registered handler callbacks for a given data.
+   */
+  private invokeHandlers(type: EventHandlerTypes, data?: unknown) {
+    const handlers = this.handlers[type];
+    if (!handlers) {
+      return;
+    }
+    handlers.forEach((h) => {
```

Comment:
This can be simplified using safe accessor
```suggestion
    handlers?.forEach((h) => {
```

Reply:
Good call, I'll apply it to my other work. This is really old code that I'm trying to get rid of some of that garbage lol.